lib/bolt/transport/local.rb

```diff
@@ -33,155 +30,140 @@ def default_input_method(executable)
-              end
-
-        yield dir
-      ensure
-        FileUtils.remove_entry dir if dir
-      end
-      private :in_tmpdir
-
-      def copy_file(source, destination)
-        FileUtils.cp_r(source, destination, remove_destination: true)
-      rescue StandardError => e
-        raise Bolt::Node::FileError.new(e.message, 'WRITE_ERROR')
-      end
-
-      def with_tmpscript(script, base)
-        in_tmpdir(base) do |dir|
-          dest = File.join(dir, File.basename(script))
-          copy_file(script, dest)
-          File.chmod(0o750, dest)
-          yield dest, dir
-        end
       end
-      private :with_tmpscript
 
-      def upload(target, source, destination, _options = {})
-        copy_file(source, destination)
-        Bolt::Result.for_upload(target, source, destination)
+      def with_connection(target)
+        conn = Shell.new(target)
+        yield conn
       end
 
-      def run_command(target, command, _options = {})
-        in_tmpdir(target.options['tmpdir']) do |dir|
-          output = @conn.execute(command, dir: dir)
-          Bolt::Result.for_command(target, output.stdout.string, output.stderr.string, output.exit_code)
+      def upload(target, source, destination, options = {})
+        with_connection(target) do |conn|
+          conn.running_as(options['_run_as']) do
+            fh = Sudoable::FileCommands.new(conn, source)
+            fh.chown(conn.run_as)
+            conn.copy_file(source, destination)
+            Bolt::Result.for_upload(target, source, destination)
+          end
         end
       end
 
-      def run_script(target, script, arguments, _options = {})
-        with_tmpscript(File.absolute_path(script), target.options['tmpdir']) do |file, dir|
-          logger.debug "Running '#{file}' with #{arguments}"
-
-          # unpack any Sensitive data AFTER we log
-          arguments = unwrap_sensitive_args(arguments)
-          if Bolt::Util.windows?
-            if Powershell.powershell_file?(file)
-              command = Powershell.run_script(arguments, file)
-              output = @conn.execute(command, dir: dir, env: "powershell.exe")
-            else
-              path, args = *Powershell.process_from_extension(file)
-              args += Powershell.escape_arguments(arguments)
-              command = args.unshift(path).join(' ')
-              output = @conn.execute(command, dir: dir)
-            end
-          else
-            if arguments.empty?
-              # We will always provide separated arguments, so work-around Open3's handling of a single
-              # argument as the entire command string for script paths containing spaces.
-              arguments = ['']
+      def run_script(target, script, arguments, options = {})
+        logger.debug "Running '#{script}' with #{arguments}"
+        if Bolt::Util.windows?
+          with_connection(target) do |conn|
+            conn.running_as(options['_run_as']) do
```

Comment:
I wouldn't expect this to work on windows. We probably need to make sure there is clean error here.

Reply:
Should we error if someone runs on winrm with `run_as`? Or just work around it and pretend it's not set?

Reply:
I think erroring is best. It may cause problems for users running `--run-as` but I think we can wait for wait for users to complain.